Added MTC with yaml description, added an SNDDesign key in run_simScr…

eduard322 opened the pull request:
1. Added first version of MTC with yaml support according to the current implementation 
2. Added SNDDesign option in run_simScript.py to choose MTC or old version (any other options can be added).

P.S. Adding a hole in the last magnet will be a separate PR, will push later the next week, since now it's not clear (for me) how to extract the absolute coordinates of the centre of the magnet before the simulation and the call of MuonShield class

olantwin commented:
pre-commit.ci autofix

olantwin commented:
Thanks @eduard322! Detailed review to come this afternoon.

olantwin commented on macro/run_simScript.py:
```diff
@@ -123,6 +123,7 @@
 )
 
 parser.add_argument("--SND", dest="SND", help="Activate SND.", action='store_true')
+parser.add_argument("--SNDDesign", dest="SNDDesign", help="Choose SND design among [1,2,3,...]. 1: old version, 2: MTC", type=int, choices=[1, 2], required=False, default=1)
```
```suggestion
parser.add_argument("--SNDDesign", help="Choose SND design among [1,2,3,...]. 1: old version, 2: MTC", type=int, choices=[1, 2], default=1)
```

I prefer not specifying keyword arguments which are left at their default values.

olantwin commented on nutaudet/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ set(INCLUDE_DIRECTORIES
 ${BASE_INCLUDE_DIRECTORIES}
 ${CMAKE_SOURCE_DIR}/shipdata
 ${CMAKE_SOURCE_DIR}/nutaudet
+${ROOT_INCLUDE_DIR}
```
Which headers is this needed for?

It would be better to be more explicit about which paths need to be included, as $ROOT_INCLUDE_DIR includes a lot of stuff.

olantwin commented on nutaudet/MTCDetector.h:
```diff
@@ -0,0 +1,92 @@
+    Int_t fLayers;
+    Double_t fZCenter;
+    Double_t fFieldY;
+    TClonesArray* fMTCDetectorPointCollection;
```
No new TClonesArrays please. This should be a vector.

eduard322 replied:
do you mean std::vector? Does it work properly with FairRoot? For example, here? (sorry, maybe stupid question)
```FairRootManager::Instance()->Register(name, title, fMTCDetectorPointCollection, kTRUE);```

olantwin replied:
Yes, and yes, FairRoot supports STL containers. I believe one change needed is using `RegisterAny` instead of `Register`. There are some examples for STL containers in the FairRoot repo.

eduard322 replied:
so far the compiler doens't like that the parent class overrides the method using TClonesArray if I use std::vector for it:
```
/afs/cern.ch/user/e/edursov/private/SIMULATIONS/my_FairShip/sw/SOURCES/FairShip/master/0/nutaudet/MTCDetector.h:44:46: error: virtual function 'GetCollection' has a different return type ('const std::vector<MTCdetPoint *> &') than the function it overrides (which has return type 'TClonesArray *')
    virtual const std::vector<MTCdetPoint*>& GetCollection(Int_t iColl) const;
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/cvmfs/ship.cern.ch/25.01/sw/slc9_x86-64/FairRoot/7b37b40c65-local1/include/FairDetector.h:63:27: note: overridden virtual function is here
    virtual TClonesArray* GetCollection(Int_t iColl) const = 0;
            ~~~~~~~~~~~~~ ^
```

Should I override it?

olantwin replied:
Ok disregard this comment for now, while I figure out the best way to proceed.
I'll probably need to discuss this with the FairRoot team, so let's not let that block your PR.

olantwin replied:
Ah, I found the example I remembered: https://github.com/FairRootGroup/FairRoot/tree/master/templates/NewDetector_stl_containers

olantwin replied:
Seems like the solution is to just return a nullptr for that method.

eduard322 replied:
I decided to finish the migration to STL since I already started, thanks for the template! Several problems:
1. I checked several options: Genie, MuonBack and PG
2. In case of absence of hits in the detector code returns error when checking the number of entries:
```
Macro finished successfully.
Output file is  ./ship.conical.MuonBack-TGeant4.root
Parameter file is  ./ship.params.conical.MuonBack-TGeant4.root
Real time  82.80425310134888  s, CPU time  76.76 s
Traceback (most recent call last):
  File "/afs/cern.ch/user/e/edursov/private/SIMULATIONS/my_FairShip/sw/slc9_x86-64/FairShip/master-local1/macro/run_simScript.py", line 599, in <module>
    if eval(x)>0: empty = False
  File "<string>", line 1, in <module>
AttributeError: 'vector<MTCdetPoint*>' object has no attribute 'GetEntries'
Exit normally
```
3. In case of presence of hits (PG) it works well, stores info, though it looks like this approach makes data less "available" for user:
```
======> EVENT:0
 MCTrack         = 58
 MCTrack.fUniqueID = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
 MCTrack.fBits   = 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648, 50331648
 MCTrack.fPdgCode = 211, 2112, -211, 2112, 2112, 2112, 211, 111, 22, -11, 22, 11, 22, 211, -211, 2112, 2112, 2112, 2112, -211
 MCTrack.fMotherId = -1, 0, 0, 2, 2, 4, 0, 0, 7, 8, 7, 10, 11, 0, 13, 14, 14, 14, 17, 0
 MCTrack.fPx     = -0, 0.277258, 0.296058, 0.206439, -0.321695, -0.329585, -0.152408, -0.322916, -0.103924, -0.0580356, -0.218993, -0.211661, -0.0704185, 0.115075, 0.230572, 0.119515, 0.0623963, -0.0468298, -0.223509, -0.183195
 MCTrack.fPy     = 0, 0.0443683, 0.190928, 0.511038, -0.488392, -0.462591, -0.160555, 0.1223, 0.104799, 0.0583438, 0.0175015, 0.017423, -0.00355552, -0.0852531, -0.158691, 0.120599, 0.547162, -0.588835, -0.424398, 0.0132632
 MCTrack.fPz     = 10, 0.506084, 0.408655, 0.0987607, 0.342641, -0.0987428, -0.25751, 0.343324, 0.104799, 0.0580988, 0.238525, 0.230947, 0.0772788, 0.819524, 0.474238, 0.468719, 0.165166, -0.0555165, -0.161583, 0.641212
 MCTrack.fM      = 0.13957, 0.939565, 0.13957, 0.939565, 0.939565, 0.939565, 0.13957, 0.134977, 0, 0.000510999, -nan, 0.000510999, -nan, 0.13957, 0.13957, 0.939565, 0.939565, 0.939565, 0.939565, 0.13957
 MCTrack.fStartX = 0, -0.0844062, -0.0844062, 7.0006, 7.0006, -2.64886, -0.0844062, -0.0844062, -0.0844063, -4.86906, -0.0844063, -2.6806, -3.40265, -0.0844062, 0.18425, 1.26627, 1.26627, 1.26627, 1.20906, -0.0844062
 MCTrack.fStartY = 0, -0.0428391, -0.0428391, 3.68965, 3.68965, -7.71053, -0.0428391, -0.0428391, -0.0428391, 4.78212, -0.0428391, 0.164644, 0.19347, -0.0428391, -0.236872, -0.959657, -0.959657, -0.959657, -1.67905, -0.0428391
 MCTrack.fStartZ = -3162, -3141.12, -3141.12, -3130.44, -3130.44, -3123.82, -3141.12, -3141.12, -3141.12, -3136.29, -3141.12, -3138.29, -3137.5, -3141.12, -3139.1, -3136.69, -3136.69, -3136.69, -3136.76, -3141.12
 MCTrack.fStartT = 0, 0.696716, 0.696716, 1.16028, 1.16028, 2.09986, 0.696716, 0.696716, 0.696716, 0.974704, 0.696716, 0.824951, 0.860575, 0.696716, 0.765895, 0.860184, 0.860184, 0.860184, 0.905468, 0.696716
 MCTrack.fW      = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
 MCTrack.fProcID = 0, 23, 23, 23, 23, 23, 23, 23, 4, 5, 4, 5, 8, 23, 23, 23, 23, 23, 23, 23
 MCTrack.fNPoints = 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
............................
 MTCdetPoint     = (vector<MTCdetPoint*>*)0x403c8f0
```

so, you see, the info is hidden in the vector of MTCdetPoint. The same story when you wanna have a look at the data via TBrowser:
<img width="226" alt="image" src="https://github.com/user-attachments/assets/0e555931-d8ae-46c3-b06d-b01d7e7cfe8f" />

The problem is that this approach doesn't create the bunch of branches like it is done with TClonesArray:
Detector.XXX 
If it is possible to make the same structure with STL vectors, it would be perfect, but so far idk how to do that

4. Though I'm pretty convinced that working with std::vector is way easier than with TClonesArray.

olantwin replied:
Thanks for investigating this.

The magic word to search for is splitlevel, which does exactly what you want. Right now we set it to -1, maybe because of the TClonesArrays.

See the open issues.

I'm on holiday until the 4th.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+    return new (clref[size]) MTCdetPoint(trackID, detID, pos, mom, time, length, eLoss, pdgCode);
+}
+
+// ClassImp(MTCDetector)
```
Delete

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+{
+    TClonesArray& clref = *fMTCDetectorPointCollection;
+    Int_t size = clref.GetEntriesFast();
+    // cout << "adding hit detid " <<detID<<endl;
```
Delete.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+
+void MTCDetector::Register()
+{
+    // FairRootManager::Instance()->Register("vetoPoint", "veto",
```
Delete

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+    TString name = "MTCdetPoint";
+    TString title = "MTC";
+    FairRootManager::Instance()->Register(name, title, fMTCDetectorPointCollection, kTRUE);
+    cout << this->GetName() << ",  Register() says: registered " << name << " collection" << endl;
```
LOG(DEBUG) would be better.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+                   fLength,
+                   fELoss,
+                   pdgCode);
+            ShipStack* stack = (ShipStack*)gMC->GetStack();
```
No C-style casts, please.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+        Double_t xmean = (fPos.X() + Pos.X()) / 2.;
+        Double_t ymean = (fPos.Y() + Pos.Y()) / 2.;
+        Double_t zmean = (fPos.Z() + Pos.Z()) / 2.;
+        // cout << "Check detID: " << detID << endl;
```
Delete

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+                   fLength,
+                   fELoss,
+                   pdgCode);
+            ShipStack* stack = (ShipStack*)gMC->GetStack();
```
No C-style casts, please

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+        Double_t ymean = (fPos.Y() + Pos.Y()) / 2.;
+        Double_t zmean = (fPos.Z() + Pos.Z()) / 2.;
+        // cout << "Check detID: " << detID << endl;
+        if (detID / 10000000 == 3) {
```
It might be better to just define the position vector conditionally to avoid duplicating everything.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+    ironVol->SetTransparency(20);
+    // (Optional: attach a magnetic field with TGeoUniformMagField if needed)
+
+    // // --- SciFi Module ---
+    // TGeoVolume* sciFiModuleVol = CreateSciFiModule("MTC_sciFi", fWidth, fHeight, fSciFiThick);
+
+    // --- Scintillator Layer (blue) ---
+
+    // // --- Assemble the layers into the envelope ---
+    // for (Int_t i = 0; i < fLayers; i++) {
+    //   // Compute the center position (z) for the current module
+    //   Double_t zPos = -totalLength/2 + (i+0.5) * moduleSpacing;
+    //   // Place the Outer Iron layer (shifted down by half the SciFi+scint thickness)
+    //   envVol->AddNode(ironVol, i, new TGeoTranslation(0, 0, zPos - (fSciFiThick + fScintThick)/2));
+    //   // Place the SciFi module (fiber module)
+    //   envVol->AddNode(sciFiModuleVol, i, new TGeoTranslation(0, 0, zPos - fScintThick/2));
+    //   // Place the Scintillator layer (shifted up by half the iron thickness)
+    //   envVol->AddNode(scintVol, i, new TGeoTranslation(0, 0, zPos + fIronThick/2));
```
Remove commented code.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+    // // Ensure fibers are visible
+    // fiberVol->SetVisibility(true);
+
+    // // --- Define rotations for fibers ---
+    // // For the U fibers: rotate X by 90° then Y by +5°
+    // TGeoRotation* rotFiberU = new TGeoRotation();
+    // //
+    // rotFiberU->RotateY(fSciFiBendingAngle);
+    // rotFiberU->RotateX(90.);
+    // // For the V fibers: rotate X by 90° then Y by -5°
+    // TGeoRotation* rotFiberV = new TGeoRotation();
+    // //
+    // rotFiberV->RotateY(-fSciFiBendingAngle);
+    // rotFiberV->RotateX(90.);
+    // // --- Place fibers in the U fiber mat ---
+    // for (int layer = 0; layer < numFiberLayers; layer++) {
+    //   Double_t zCenter = -fiberMatThick/2 + (layer + 0.5) * (layerThickness + antioverlap);
+    //   for (int j = 0; j < fNumFibers; j++) {
+    //     Double_t xPos = -fSciFiActiveAreaX/2 + (j + 0.5) * fFiberPitch;
+    //     // Create a combined translation+rotation
+
+    //     TGeoCombiTrans* ctU = new TGeoCombiTrans("", xPos, 0, zCenter, rotFiberU);
+    //     sciFiLayerMotherUVol->AddNode(fiberVol, 100000000 + LayerId * 100000 + layer * fNumFibers + j, ctU);
+    //     // cout << "Define Scifi Fibre: " << 100000000 + LayerId * 100000 + layer * 10000 + j << "  " << LayerId << "
+    //     " << 0 << "   " << layer << "  " << j << endl; cout << "Define Scifi Fibre: " << 100000000 + LayerId * 100000
+    //     + layer * fNumFibers + j << "  " << LayerId << "  " << 0 << "   " << layer << "  " << j << endl;
+    //   }
+    // }
+
+    // // --- Place fibers in the V fiber mat ---
+    // for (int layer = 0; layer < numFiberLayers; layer++) {
+    //   Double_t zCenter = -fiberMatThick/2 + (layer + 0.5) * (layerThickness + antioverlap);
+    //   for (int j = 0; j < fNumFibers; j++) {
+    //     Double_t xPos = -fSciFiActiveAreaX/2 + (j + 0.5) * fFiberPitch;
+    //     TGeoCombiTrans* ctV = new TGeoCombiTrans("", xPos, 0, zCenter, rotFiberV);
+    //     sciFiLayerMotherVVol->AddNode(fiberVol, 200000000 + LayerId * 100000 + layer * fNumFibers + j, ctV);
+    //     // cout << "Define Scifi Fibre: " << 200000000 + LayerId * 100000 + layer * 10000 + j << "  " << LayerId << "
+    //     " << 1 << "   " << layer << "  " << j << endl;
+    //   }
+    // }
```
Delete commented code.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+            Double_t y = -height / 2 + cellSizeY * (j + 0.5);
+            // 200000000 + LayerId * 100000 + layer * 10000 + j
+            motherVol->AddNode(cellVol, 30000000 + LayerId * 100000 + i * nY + j, new TGeoTranslation(x, y, 0));
+            // cout << "Define Scint cell: " << 200000000 + LayerId * 100000 + i*nY+j << "  " << LayerId << "  " <<
+            // i*nY+j << endl;
```
Remove

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+                                              Double_t transparency,
+                                              Int_t LayerId)
+{
+    TGeoBBox* mother = new TGeoBBox(Form("%s_mother", name), width / 2, height / 2, thickness / 2);
```
Consider using auto for these kind of lines where we repeat the type.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+//     // Mother volume for the SciFi module
+//     TGeoBBox* modMother = new TGeoBBox(Form("%s_mother", name), width/2, height/2, thickness/2);
+//     // We use the SciFi material for the mother; adjust if you have a dedicated one
+//     TGeoVolume* modMotherVol = new TGeoVolume(Form("%s_mother", name), modMother,
+//     gGeoManager->GetMedium("SciFiMat")); modMotherVol->SetLineColor(kGreen+2); modMotherVol->SetTransparency(40);
+
+//     // --- Lower Internal Iron ---
+//     TGeoBBox* lowerIronBox = new TGeoBBox(Form("%s_lowerIron", name), width/2, height/2, lowerIronThick/2);
+//     TGeoVolume* lowerIronVol = new TGeoVolume(Form("%s_lowerIron", name), lowerIronBox,
+//     gGeoManager->GetMedium("iron")); lowerIronVol->SetLineColor(kGray+1); lowerIronVol->SetTransparency(20);
+//     // Position: at the bottom of the module
+//     modMotherVol->AddNode(lowerIronVol, 1, new TGeoTranslation(0, 0, -thickness/2 + lowerIronThick/2));
+
+//     // --- Fiber Mat U ---
+//     TGeoBBox* fiberMatBoxU = new TGeoBBox(Form("%s_fiberMat_U", name), width/2, height/2, fiberMatThick/2);
+//     TGeoVolume* fiberMatVolU = new TGeoVolume(Form("%s_fiberMat_U", name), fiberMatBoxU,
+//     gGeoManager->GetMedium("SciFiMat")); fiberMatVolU->SetLineColor(kYellow); fiberMatVolU->SetTransparency(30);
+//     // Position: above lower iron
+//     modMotherVol->AddNode(fiberMatVolU, 1, new TGeoTranslation(0, 0, -thickness/2 + lowerIronThick +
+//     fiberMatThick/2));
+
+//     // --- Fiber Mat V ---
+//     TGeoBBox* fiberMatBoxV = new TGeoBBox(Form("%s_fiberMat_V", name), width/2, height/2, fiberMatThick/2);
+//     TGeoVolume* fiberMatVolV = new TGeoVolume(Form("%s_fiberMat_V", name), fiberMatBoxV,
+//     gGeoManager->GetMedium("SciFiMat")); fiberMatVolV->SetLineColor(kYellow); fiberMatVolV->SetTransparency(30);
+//     // Position: above Fiber Mat U plus air gap
+//     modMotherVol->AddNode(fiberMatVolV, 1, new TGeoTranslation(0, 0, -thickness/2 + lowerIronThick + fiberMatThick +
+//     airGap + fiberMatThick/2));
+
+//     // --- Upper Internal Iron ---
+//     TGeoBBox* upperIronBox = new TGeoBBox(Form("%s_upperIron", name), width/2, height/2, upperIronThick/2);
+//     TGeoVolume* upperIronVol = new TGeoVolume(Form("%s_upperIron", name), upperIronBox,
+//     gGeoManager->GetMedium("iron")); upperIronVol->SetLineColor(kGray+1); upperIronVol->SetTransparency(20);
+//     // Position: at the top of the module
+//     modMotherVol->AddNode(upperIronVol, 1, new TGeoTranslation(0, 0, thickness/2 - upperIronThick/2));
+
+//     // Optionally, you can add fiber placements inside the fiber mats here.
+
+//     return modMotherVol;
+//   }
```
Delete.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+        return -1111;
+    }
+    TGeoMedium* medium = gGeoManager->GetMedium(name);
+    if (medium != NULL)
```
Prefer using nullptr.

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,543 @@
+    FairGeoMedium* ShipMedium = media->getMedium(name);
+
+    if (!ShipMedium) {
+        Fatal("InitMedium", "Material %s not defined in media file.", name);
```
Prefer using LOG(FATAL)

olantwin reviewed (changes requested):
Looks good generally.

Some general things to consider:
- It might be better to make a new directory for the next generation of the SND, as it's not exclusively a $\nu_\tau$ detector.
- We might want to use a namespace for new classes. I'd suggest SHiP::SND, maybe even SHiP::snd::mtc, so that the actual class names can be nice and short.

Sorry for being stricter than usual, but for new code I would like to apply higher standards than we can enforce for the legacy code.

Concerning the position for the integration, I think @Gfrisella should be able to provide you a way to calculate the z-position of the last magnet given the information available in `geometry_config.py`.

antonioiuliano2 commented on geometry/geometry_config.py:
```diff
@@ -70,11 +70,15 @@
 if "SND" not in globals():
     SND = True
+if "SNDDesign" not in globals():
+    SNDDesign = 1
```
What is the difference between this and the previous NuTauTargetDesign flag?
I would avoid duplicates since it creates confusion, so if we keep this one we should remove the old one.

olantwin replied:
I'd vote in favour of removing NuTauTargetDesign:
- The SND is **not** exclusively a $\nu_\tau$ detector, and one shouldn't need to have studied the pre-ECN3 history to use the script
- The "Target" part might cause confusion. Not everyone might expect the scattering detector be called target, even if if technically is a secondary target for BSM and neutrinos.

Maybe `SND_design` or `SND_DESIGN` would be better than `SNDDesign` to follow common python naming conventions.

eduard322 commented:
I applied all the changes except for STL vectors, because they are not compatible at all with the whole fairship, so it will be better to make another PR, where all the TClonesArray are replaced with std::vector and all the other things regarding this global change are fixed.

olantwin commented:
> I applied all the changes except for STL vectors, because they are not compatible at all with the whole fairship, so it will be better to make another PR, where all the TClonesArray are replaced with std::vector and all the other things regarding this global change are fixed.

Ok, thanks again for investigating.

Maybe we can discuss your findings about STL vectors at today's computing meeting and add the information to #430 ?

I'll review the other changes again now.

eduard322 commented:
> Maybe we can discuss your findings about STL vectors at today's computing meeting and add the information to #430 ?

Sure, I have also some more things to discuss for the computing meeting :)

olantwin commented:
> Sure, I have also some more things to discuss for the computing meeting :)

Great! Shall I add an agenda point for you?

eduard322 commented:
> > Sure, I have also some more things to discuss for the computing meeting :)
> 
> Great! Shall I add an agenda point for you?

yep

olantwin commented:
> > > Sure, I have also some more things to discuss for the computing meeting :)
> > 
> > 
> > Great! Shall I add an agenda point for you?
> 
> yep

Done

olantwin commented on nutaudet/MTCDetector.cxx:
```diff
@@ -0,0 +1,418 @@
+void MTCDetector::Register(){
+    FairRootManager::Instance()->RegisterAny("MTCdetPoint", fMTCPointPtr, kTRUE);
+    cout << this->GetName() << ",  Register() says: registered " << "MTCdetPoint" <<" collection"<<endl;
+  }
```
Remove this brace, it causes the compilation to fail.

olantwin replied:
Fixed, now it should at least build...

olantwin reviewed (changes requested):
Please run clang-format over the new files and the modified lines.

The yaml and modifications to the old files look fine, I'll need a bit more time to read the new C++.

olantwin commented:
This PR will be reviewed when it at least builds. Currently it's a mix of TClonesArray and STL code mixed all together resulting in one compilation error after another.

eduard322 commented:
Sorry for the previous commits, I was confused by git(hub) and messed up. I added some updates in the following commits:
1. MTC is now in the separate folder
2. MTC is now magnetized (somehow I forgot to add the field)
3. MTC now can be automatically placed in the center of the last magnet. To do that, I used python/ShipUtils.py script, I modified it, fixed some syntax (it still backward compatible with the only script where it was used).

Minor updates:
I modified paths in the shipDet_conf.py so that it is done now via os.path.join() (hope it's that minor to not put it into different PR).

olantwin commented:
pre-commit.ci autofix

olantwin commented:
PS: I'd recommend setting up whatever editor you use to show trailing whitespace and write the end of files properly (see the pre-commit lints that keep failing).

olantwin reviewed (approved):
Looks good to me. Could you please add a changelog entry?

Do we want to have this as several commits or squash everything together into one?

eduard322 commented:
> This is with a constant field for now, right?
so far yes, one can change the field strength in the yaml file or set it to 0.

> Great! Is this something that can be easily copied by e.g. @amagnan for the other parts of the SND?

yep, the ShipUtils.py script baically calculates the coordinates of all the magnets, I use it at the shipDet_conf.py level.

eduard322 commented:
> Looks good to me. Could you please add a changelog entry?
Done
> Do we want to have this as several commits or squash everything together into one?
I guess one can squash them into one to make the updates less spread
